File: url_loader.py
import csv
import os
from urllib.parse import urlparse


class URLLoader:
    def __init__(self):
        self.urls = []
    
    def load_urls_from_csv(self, file_path):
        if not file_path:
            raise ValueError("CSV file path is required")
        
        if not os.path.exists(file_path):
            raise FileNotFoundError(f"CSV file not found: {file_path}")
        
        if not file_path.lower().endswith('.csv'):
            raise ValueError(f"Invalid file type. Expected .csv, got {os.path.splitext(file_path)[1]}")
        
        self.urls = []
        results = []
        
        try:
            with open(file_path, 'r', encoding='utf-8') as file:
                reader = csv.DictReader(file)
                for row in reader:
                    for value in row.values():
                        if self.is_valid_url(value):
                            results.append(value)
            
            self.urls = list(set(results))
            print(f"Loaded {len(self.urls)} unique URLs from {file_path}")
            return self.urls
            
        except Exception as e:
            raise Exception(f"Error reading CSV file: {str(e)}")
    
    def is_valid_url(self, string):
        try:
            result = urlparse(string)
            return result.scheme in ['http', 'https'] and result.netloc
        except:
            return False
    
    def get_urls(self):
        return self.urls
    
    def get_url_count(self):
        return len(self.urls)
    
    def clear_urls(self):
        self.urls = []

File: url-loader.js
const fs = require('fs');
const path = require('path');
const csv = require('csv-parser');

class URLLoader {
  constructor() {
    this.urls = [];
  }

  async loadUrlsFromCSV(filePath) {
    return new Promise((resolve, reject) => {
      if (!filePath) {
        reject(new Error('CSV file path is required'));
        return;
      }

      if (!fs.existsSync(filePath)) {
        reject(new Error(`CSV file not found: ${filePath}`));
        return;
      }

      const extension = path.extname(filePath).toLowerCase();
      if (extension !== '.csv') {
        reject(new Error(`Invalid file type. Expected .csv, got ${extension}`));
        return;
      }

      this.urls = [];
      const results = [];

      fs.createReadStream(filePath)
        .pipe(csv())
        .on('data', (row) => {
          Object.values(row).forEach(value => {
            if (this.isValidURL(value)) {
              results.push(value);
            }
          });
        })
        .on('end', () => {
          this.urls = [...new Set(results)];
          console.log(`Loaded ${this.urls.length} unique URLs from ${filePath}`);
          resolve(this.urls);
        })
        .on('error', (error) => {
          reject(new Error(`Error reading CSV file: ${error.message}`));
        });
    });
  }

  isValidURL(string) {
    try {
      const url = new URL(string);
      return url.protocol === 'http:' || url.protocol === 'https:';
    } catch {
      return false;
    }
  }

  getUrls() {
    return this.urls;
  }

  getUrlCount() {
    return this.urls.length;
  }

  clearUrls() {
    this.urls = [];
  }
}

module.exports = URLLoader;